File: ios/MRT/Core/Network/ProtobufCodec.swift
import Foundation
import SwiftProtobuf

enum ProtobufCodecError: Error, Equatable {
    case frameTooShort
    case invalidLengthPrefix(expected: Int, actual: Int)
}

enum ProtobufCodec {
    static func encode(_ envelope: Mrt_Envelope) throws -> Data {
        let payload = try envelope.serializedData()
        var length = UInt32(payload.count).bigEndian
        var framed = Data(bytes: &length, count: MemoryLayout<UInt32>.size)
        framed.append(payload)
        return framed
    }

    static func decode(_ data: Data) throws -> Mrt_Envelope {
        guard data.count >= 4 else {
            throw ProtobufCodecError.frameTooShort
        }

        let declaredLength = data.prefix(4).reduce(0) { ($0 << 8) | UInt32($1) }
        let payload = data.dropFirst(4)

        guard payload.count == Int(declaredLength) else {
            throw ProtobufCodecError.invalidLengthPrefix(
                expected: Int(declaredLength),
                actual: payload.count
            )
        }

        return try Mrt_Envelope(serializedBytes: payload)
    }
}

File: android/app/src/main/java/com/mrt/app/core/network/ProtobufCodec.kt
package com.mrt.app.core.network

import mrt.Mrt
import java.nio.ByteBuffer
import java.nio.ByteOrder

sealed class ProtobufCodecError(message: String) : IllegalArgumentException(message) {
    data object FrameTooShort : ProtobufCodecError("Frame shorter than 4-byte length prefix")
    data class InvalidLengthPrefix(val expected: Int, val actual: Int) :
        ProtobufCodecError("Invalid length prefix. Expected $expected bytes, got $actual")
}

object ProtobufCodec {
    fun encode(envelope: Mrt.Envelope): ByteArray {
        val payload = envelope.toByteArray()
        return ByteBuffer
            .allocate(4 + payload.size)
            .order(ByteOrder.BIG_ENDIAN)
            .putInt(payload.size)
            .put(payload)
            .array()
    }

    fun decode(data: ByteArray): Mrt.Envelope {
        if (data.size < 4) {
            throw ProtobufCodecError.FrameTooShort
        }

        val payloadLength = ByteBuffer
            .wrap(data, 0, 4)
            .order(ByteOrder.BIG_ENDIAN)
            .int
        val payload = data.copyOfRange(4, data.size)

        if (payload.size != payloadLength) {
            throw ProtobufCodecError.InvalidLengthPrefix(
                expected = payloadLength,
                actual = payload.size,
            )
        }

        return Mrt.Envelope.parseFrom(payload)
    }
}
